Previous Loot Optimisation

adverserath opened the pull request:
* CalculateValue method will now factor in players luck to increase the modifiers on items, instead of every 600 points. This removes the constant rerolling for better loot and just scales the property weight up from the first roll based on luck.
* Players with 0 luck get no benefit, but a player with 60 luck would get a 1% increase, this is a steady increase 10% at 600 luck, 20% at 1800 luck. Meaning higher luck WILL result in higher property weights on items.

* Treasure Chests now send the cartographers mobile into the loot generation, so treasure hunters will get better loot with luck
* In loot generation the percLow value was being ignored, and setting itself to 0. I've set it to not ignore, but reduced the low property to scale against the luck calculation.

* This is the significant as instead of properties always rolling from 0 > 80, or 0>30 they will always roll at (80/3)>80 (at best, becuase the minimum value is multiplied against the perc from the luck calculation). This again means higher luck will also give higher chances to roll against the minimum values

adverserath committed:
Loot Optimisation
* Budget Calculator optimised to roll higher values as luck increases
* Runic scaling is applied to all loot and Mod scale perclow calculated off perc tied directly to luck
This means 20% higher values on average, but can roll lesser than before to balance

adverserath committed:
Use Convert.ToInt over casting due to loss of precision

adverserath committed:
* Add treasure chest finder into runic calculation

TrueUO committed:
Merge branch 'master' into MobLootGeneration

TrueUO committed:
Merge branch 'master' into MobLootGeneration

TrueUO committed:
Merge branch 'master' into MobLootGeneration

TrueUO committed:
Merge branch 'master' into MobLootGeneration

TrueUO committed:
Merge branch 'master' into MobLootGeneration

TrueUO committed:
Merge branch 'master' into MobLootGeneration

TrueUO committed:
Merge branch 'master' into MobLootGeneration

adverserath commented on Scripts/Services/TreasureMaps/TreasureMapInfo.cs:
```diff
@@ -1022,7 +1022,7 @@ public static void Fill(Mobile from, TreasureMapChest chest, TreasureMap tMap)
 
                 if (item != null)
                 {
-                    RunicReforging.GenerateRandomItem(item, from is PlayerMobile pm ? pm.RealLuck : from.Luck, min, max, chest.Map);
```
Add player who found the chest to loot generation

adverserath commented on Scripts/Services/TreasureMaps/TreasureMapChest.cs:
```diff
@@ -238,7 +238,7 @@ public static void Fill(Mobile from, LockableContainer cont, int level, bool isS
                         int min, max;
                         GetRandomItemStat(out min, out max, propsScale);
 
```
Add player who found the chest to loot generation (covers SOS too)

adverserath commented on Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs:
```diff
@@ -941,30 +941,12 @@ public static bool ApplyResistance(Item item, int value, AosElementAttribute att
-                percent = Utility.RandomMinMax(0, perchigh);
-            }
-
+            decimal percent = Utility.RandomMinMax(perclow, perchigh);
```
A minimum value is always calculated on loot, but was ignored. So the low is now being used again here, to stop everything rolling from 0 upwards

adverserath commented on Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs:
```diff
@@ -941,30 +941,12 @@ public static bool ApplyResistance(Item item, int value, AosElementAttribute att
-            percent *= 10000 + scaledBy;
-
-            return min + (max - min) * percent / 1000001;
+            return Convert.ToInt32(Math.Round(((max - min) * percent / 100) + min));
```
Same calculation as previous, but in less code

TrueUO replied:
This is great. Anything that makes it easier to read / less code.

adverserath commented on Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs:
```diff
@@ -1872,13 +1854,14 @@ public static bool GenerateRandomItem(Item item, Mobile killer, BaseCreature cre
             {
                 int budget = Utility.RandomMinMax(minBudget, maxBudget);
-                GenerateRandomItem(item, null, budget, LootPack.GetLuckChance(luck), ReforgedPrefix.None, ReforgedSuffix.None, map);
+                GenerateRandomItem(item, finder, budget, LootPack.GetLuckChance(luck), ReforgedPrefix.None, ReforgedSuffix.None, map);
```
allow treasure chests to pass in mobile (the finder)

TrueUO replied:
Does production UO take into account t-hunters luck? Just wondering.

adverserath commented on Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs:
```diff
@@ -1938,18 +1921,23 @@ public static void GenerateRandomItem(Item item, Mobile killer, int basebudget,
 
-                    double perc = 0.0;
-                    double highest = 0.0;
+                    int luckyRandom = 10000;
```
Main part of the change:
before: the perc is calculated by constantly rerolling and replacing if its a higher value
now: the perc is rolled once, but then rerolled against the previous number for each 600. then any remaining luck is calculated in too
example: with 900 luck
1st roll as normal:                     luckyRandom =  random(1,10000)=3700
2nd roll for 600 luck :                luckyRandom = random(3700,10000)=6000
3rd roll for 300 remaining luck:  6000+ (random(6000,10000)-6000)*0.5  = 7000

adverserath commented on Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs:
```diff
@@ -1977,12 +1965,13 @@ public static void GenerateRandomItem(Item item, Mobile killer, int basebudget,
-                        perchigh = Math.Max(50, Math.Min(500, budget) / mods);
-                        perclow = Math.Max(20, perchigh / 3);
+                        perchigh = Math.Max(50, Math.Min(550, budget) / mods);
+                        perclow = Math.Max(20, Convert.ToInt32(perchigh / 3 * perc));
```
the perclow was always ignored, so I've scaled it down based on the perc rolled at the start.
Higher luck will give higher minimums, but will always be a maximum of 33% of the perchigh if perc was 1.0, if perc rolled 0% then you would still roll 0 to perchigh as previously. But luck always matters now.

adverserath commented on Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs:
```diff
@@ -1992,7 +1981,7 @@ public static void GenerateRandomItem(Item item, Mobile killer, int basebudget,
 
                         perchigh = 100;
-                        perclow = Utility.RandomMinMax(50, 70);
+                        perclow = Convert.ToInt32(Utility.RandomMinMax(50, 70) * perc);
```
Same as above, this 50-70 was always ignored. So its being reduced against the perc rolled at the start. We would see less Legendary artifacts with 1% properties and more higher values. But 1% is still possible on bad real life luck.

TrueUO committed:
Merge branch 'master' into MobLootGeneration

TrueUO committed:
Merge branch 'master' into MobLootGeneration

TrueUO committed:
Merge branch 'master' into MobLootGeneration

adverserath committed:
Allow overcapped low properties on high fame mobs

adverserath committed:
Merge branch 'MobLootGeneration' of https://github.com/TrueUO/TrueUO into MobLootGeneration

adverserath committed:
Allow max property override for Treasure/SOS
* SOS 2/3 + 1 Max Prop
* SOS 4 +2 Max Prop
* Stash/Supply + 0 Max Prop
* Cache/Horde +1 Max Prop
* Trove + 2 Max Prop

adverserath committed:
Additional Props
*check in broken line

adverserath committed:
Loot Modification
* additional props now processed correctly

adverserath committed:
Loot Fixes and Changes
-Paragon chests will now calculate against the killers luck
- Fix "Blackthorn invasion loot" now uses luckchance not luck
- Fix "Blackthorn points loot" now uses awarding player luck, not the last monster top damager luck
- Fix "Shrine Battle loot" now uses luckchance not luck
- Fix "Treasures of Doom" now uses awarding player luck, not the last monster top damager luck
- Khaldun chest now Fills when unlocked against t-hunters luck
- Fix "Khaldun Points" now uses awarding player luck, not the last monster top damager luck
- Regal chest now Fills when unlocked against t-hunters luck
- Fix "Kotl city points" now uses awarding player luck, not the last monster top damager luck
- Fix "Sorcerers Dungeon" now uses awarding player luck, not the last monster top damager luck

TrueUO committed:
Merge branch 'master' into MobLootGeneration

TrueUO committed:
Merge branch 'master' into MobLootGeneration

adverserath committed:
Chest Fix
* Assign the finder to the loot generation

adverserath committed:
Merge branch 'MobLootGeneration' of https://github.com/TrueUO/TrueUO into MobLootGeneration

TrueUO committed:
Merge branch 'master' into MobLootGeneration

DocKnight86 committed:
Merge branch 'master' into MobLootGeneration

TrueUO committed:
Merge branch 'master' into MobLootGeneration

TrueUO committed:
Merge branch 'master' into MobLootGeneration

adverserath committed:
* Spread loot against mob fame and luck

adverserath committed:
Merge branch 'MobLootGeneration' of https://github.com/TrueUO/TrueUO into MobLootGeneration

adverserath committed:
* Instanced loot uses each players luck
* Distribute loot by damage dealt

adverserath committed:
* Remove previously added event from commit

adverserath committed:
Revert "* Remove previously added event from commit"

This reverts commit 482edb487949314ea5e50109656908bbbd0e051d.

adverserath committed:
Revert "* Instanced loot uses each players luck"

This reverts commit da2588cd82fbf88bc7b9f2111db1e74d6c2c4241.
